packages/cli/src/auth/index.ts

```diff
@@ -43,98 +32,170 @@ export function generatePKCE () {
-
-        const error = responseParams.get('error')
-        const errorDescription = responseParams.get('error_description')
-
-        if (code && state === codeVerifier) {
-          res.write(`
-      <html>
-      <body>
-        <div style="height:100%;width:100%;inset:0;position:absolute;display:grid;place-items:center;background-color:#EFF2F7;text-align:center;font-family:Inter;">
-          <h3 style="font-weight: 200;"><strong style="color:#45C8F1;">@checkly/cli</strong> login success! </br></br>You can go back to your terminal</br></br>This window should close itself in 3 seconds.</h3>
-        </div>
-        <script>setTimeout(function() {window.close()}, 3000);</script>
-      </body>
-      </html>
-    `)
-          resolve(code)
-        } else {
-          res.write(`
-      <html>
-      <body>
-        <div style="height:100%;width:100%;inset:0;position:absolute;display:grid;place-items:center;background-color:#EFF2F7;text-align:center;font-family:Inter;">
-          <h3 style="font-weight:200;">Login failed, please try again!</h3>
-          <p>
-            <b>${error}</b>: ${errorDescription}
-          </p>
-        </div>
-      </body>
-      </html>
-    `)
-        }
+export class AuthContext {
+  authenticationUrl: string
 
-        res.end()
-      }
-    })
+  #codeChallenge: string
+  #codeVerifier: string
 
-    server.listen(4242).on('error', (err: any) => {
-      if (err.code === 'EADDRINUSE') {
-        reject(new Error('Unable to start a local server on port 4242.' +
-          ' Please check that `checkly login` isn\'t already running in a separate tab.' +
-          ' On OS X and Linux, you can run `lsof -i :4242` to see which process is blocking the port.'))
-      } else {
-        reject(err)
-      }
-    })
-  })
-}
+  #axiosInstance!: Axios
+  #accessToken!: string
+
+  constructor () {
+    const { codeChallenge, codeVerifier } = generatePKCE()
+    this.#codeChallenge = codeChallenge
+    this.#codeVerifier = codeVerifier
+
+    this.authenticationUrl = this.#generateAuthenticationUrl()
+  }
 
-export async function getAccessToken (code: string, codeVerifier: string) {
-  const tokenParams = new URLSearchParams({
-    grant_type: 'authorization_code',
-    client_id: AUTH0_CLIENT_ID,
-    code_verifier: codeVerifier,
-    code,
-    redirect_uri: 'http://localhost:4242',
-  })
-
-  const tokenResponse = await axios.post(
-    'https://auth.checklyhq.com/oauth/token',
-    tokenParams,
-    {
+  async getAuth0Credentials () {
+    const { access_token: accessToken, id_token: idToken } = await this.#getAccessToken()
+
+    this.#accessToken = accessToken
+    this.#axiosInstance = axios.create({
+      baseURL: getApiDefaults().baseURL,
       headers: {
-        'Content-Type': 'application/x-www-form-urlencoded',
-        'Accept-Encoding': '*',
+        Accept: 'application/json, text/plain, */*',
+        Authorization: `Bearer ${this.#accessToken}`,
       },
-    },
-  )
+    })
```

Comment:
I have 2 problems with this:
1. Assigning values to the object in between steps usually leads to null|undefined errors
2. You are assigning values in a `get` method which hides the axios creation

Reply:
Why don't we just pass axios as function paramters? This would be avoided then